back/src/forms/mutations/mark-as.ts

```diff
@@ -62,11 +68,27 @@ export function markAsProcessed(
   { id, processedInfo }: MutationMarkAsProcessedArgs,
   context: GraphQLContext
 ): Promise<Form> {
+  const operation = PROCESSING_OPERATIONS.find(
+    otherOperation =>
+      otherOperation.code === processedInfo.processingOperationDone
+  );
+
+  if (operation == null) {
+    throw new ValidationError(
```

Comment:
Je serai curieux de connaître votre retour d'expérience sur `graphql-shield`. Sur le papier ça a l'air intéressant mais je trouve que dans la pratique ça pose plusieurs problèmes :

- Ça découple du code qui est indissociable. Pour que ces fonctions soient appelés (par exemple `markAsProcessed`), la validation des données en entrée est inévitable. Le soucis c'est que `graphql-shield` encourage cette validation complètement ailleurs et de façon totalement découplé : on peut très bien modifier la validation sans que ça n'est d'impact direct sur le code de la fonction, et inversement. Alors TypeScript pourrait nous prévenir si il y avait une incohérence dans les arguments et/ou on s'en apercevrait beaucoup plus facilement si tout était au même endroit.
- La validation dans ces fonctions est de toute façon inévitable. Par exemple, si il faut récupérer un `Form` depuis la base de donnée mais que pour une raison X ou Y la base de donnée retourne `null`, il faut le prendre en compte.
- Ça peut encourage à prendre des partis pris, du genre partir du principe que l'ID du formulaire est valide puisqu'il est passé au travers du "shield". Ce qui peut pousser à écrire du code pas très bien typé, fragile et difficile à tester unitairement.

J'ai l'impression que la base de code gagnerait en maintenabilité et en simplicité si la validation était faite par le principal concerné. Qu'en pensez-vous ?

Reply:
On a eu une discussion très intéressante avec Laurent 🙌 On est d'accord de faire les vérifications de ce type quand c'est nécessaire, c'était mon point principal. Du reste, `graphql-shield` fait le job et les tests d'intégration couvriront l'ensemble 👌